### ISSUE TITLE
Typing Return of Function is not checking returned variable types properly

### ISSUE THREAD

**Issue by chenxeed** [microsoft/TypeScript#52823]
# Bug Report

Return types of function seems not checking the returned variable if it match the explicit return type.

### 🕗 Version & Regression Information

Tested the code below on http://www.typescriptlang.org/play/?ts=Nightly and the problem still occurs.

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.0.0-dev.20230216#code/JYOwLgpgTgZghgYwgAgKoGdrIN4ChkHICumUAXMumFKAOYDc+hUA9gDYQVU0gO4C+uXAhYgqxUgHVgYABYAFOOnTIAvDiYES0CgHJZcA7oA0m5Kw57ZEaybMAHJegDuLKABM9AWxY-dAoQB6QOQAcQB5cIARClDgADcUABUAZWRoVihkJXSAD3sIBEh3YVFxBGsEAGsIdwwsdQAKAEoKeqzVAD4NQnMIMCIoEB7e5nZOZH0INhY7UcJtckmDI1N5gkdlVw9vOC84f17BQVxg5AAhAEEY5ABJMGR3Fgh0EF0H2gSUORQMt1KxA8iCAKoUanVSGpkC02pCuiMCFB+oNhotpHJFMoAgDxCAIM52gBGKGg6q1dotXB4gmkQkAOkWVPxRLpFggQhEgOQ1PaACYocDSeCKc0mTToLyGaQxXzWeMZaRJZsXG53EA

### 💻 Code

```ts
interface User {
    user: string;
    role: string;
}

const userWithPass = {
    user: 'haha',
    role: 'hehe',
    password: 'momo'
}

// GOOD: Give TS error as expected
const checkedUser = (): User => {
    return {
        role: 'helo',
        user: 'haha',
        password: 'mama'
    }
}

// BAD: It doesn't give the error
const uncheckedUser = (): User => {
    return userWithPass
}
```

### 🙁 Actual behavior

On `uncheckedUser`, no error defined despite it has extra `password` on the object.

### 🙂 Expected behavior

`uncheckedUser` should show the same typing error with `checkedUser` which is:

```
Type '{ role: string; user: string; password: string; }' is not assignable to type 'User'.
  Object literal may only specify known properties, and 'password' does not exist in type 'User'.
```

**Comment by fatcerberus** [comment]
> **Object literal** may only specify known properties, and 'password' does not exist in type 'User'.

Emphasis mine.  This check is only done for fresh object literals assigned directly to a type; otherwise, the type with extra properties is considered a valid subtype.  To prevent that, you would need exact types, which TS doesn't support. 
 See https://github.com/microsoft/TypeScript/issues/12936.

**Comment by fatcerberus** [comment]
To TypeScript team: Can we have a FAQ entry explaining that EPC is not a normal part of type checking?  This gets reported often enough that it should probably be mentioned under https://github.com/microsoft/TypeScript/wiki/FAQ#common-bugs-that-arent-bugs

**Comment by MauroTaliente** [comment]
const userWithPass = {
    user: 'haha',
    role: 'hehe',
    password: 'momo'
} as const;
